### ISSUE TITLE
Table of contents

### ISSUE THREAD

**Issue by klaascuvelier** [toddmotto/linkjuice#8]
Hi @toddmotto, this a great plugin.

I was wondering if you ever got the request to get a list of generated links.
I use this plugin myself and wanted to have some sort of "table of contents" dynamically generated based upon the links that were generated by your plugin.

I achieved it this way now:

``` js
var linkNodes = [].slice.call(document.getElementsByClassName(linkjuiceClass));
var menu = document.getElementById('table-of-contents');

menu.innerHTML = linkNodes
    .map(function (node) {
        return '<li><a href="' + node.href + '">' + node.innerHTML + '</a></li>';
    })
    .join('');
```

This works, but is obviously not the most desirable way of doing it.
I could also populate an array using the `contentFn` function, but that's very impure.

An alternative would be to give the list of links generated (`id` + `title`) as function return or have something link `linkjuice.links` (Array/NodeList) which can be iterated on.

I can help you implement this if you'd want that, but if I am the only person who ever had this request, you can just ignore this :)

**Comment by toddmotto** [comment]
Hey man, this is an awesome idea. Was actually speaking to colleague of mine about this a few weeks back. Simple flag/element selector with a new `{ tableOfContents: '.my-class' }`? With a `false` default?

**Comment by klaascuvelier** [comment]
I'm not sure if I completed understand your suggestion: 

You want to add an option `tableOfContents`? And when it's a string (not `false`) add that class onto the wrapped element?  So the developer would still have to do `document.querySelectorAll('`table-of-contents-class`')` himself/herself to create the actual table.

Is that correct?

**Comment by toddmotto** [comment]
Well, people might want to use it without generating a TOC. So the idea is they'd do (at the top of their document):

``` html
<!-- blog title etc -->
<div class="toc"></div>
<!-- blog content -->
```

And then the `init`:

``` js
linkjuice.init({
  tableOfContents: '.toc'
});
```

This would automagically compile + inject the list for them at runtime :) that was my idea anyway. And if they don't even have a TOC, they just won't use the option.

**Comment by klaascuvelier** [comment]
Makes sense. I assume this would also require `contentFn` for the TOC?
So 2 new options: `tableOfContents` + `tableOfContentsContentFn`?

**Comment by klaascuvelier** [comment]
Follow up question(s):

If the selectors are `h1, h2, h3, ...`, do you nest lists in the table of contents, or would you keep it flat?
Do you make nesting an option?
The `tableOfContentsContentFn` could also have a 2nd argument which is `level`, or let the user detect the level through `node.tag`. But that makes it semantically incorrect.

**Comment by toddmotto** [comment]
Haha, this was what I was thinking also for nested TOCs. Maybe we could somehow implement header mapping yeah.

**Comment by toddmotto** [comment]
Thoughts on this?

``` js
linkjuice.init({
  tableOfContents: {
    nested: true, // boolean: creates the nested TOC
    selector: '.toc', // string: element Node to inject into
    contentFn() {...} // function: customise template + injected content
  }
});
```

**Comment by klaascuvelier** [comment]
Yeah that sounds good. Keeps the config quite simple and gives end user a lot of control.
Is the `contentFn` used for the "title" node only and the plugin takes care of `ul` and `li` elements?

**Comment by toddmotto** [comment]
Yeah, I think we can detect nested of headings - this may need to become a configurable option though to tell it what headings to map to nested lists?

``` js
linkjuice.init({
  tableOfContents: {
    nested: ['h1', 'h2'], // array: creates the nested TOC, nests based off new heading selectors?
    selector: '.toc', // string: element Node to inject into
    contentFn() {...} // function: customise template + injected content
  }
});
```

**Comment by toddmotto** [comment]
Maybe if you use `nested: true`, the TOC will automatically be generated from h1/h2/h3 ? Might be nice feature.

**Comment by klaascuvelier** [comment]
Yeah, that makes sense to me.

Btw, the [Contribution](https://github.com/toddmotto/linkjuice#contributing) section mentions unit test. You want me to set up some Karma infrastructure for this as well?

**Comment by klaascuvelier** [comment]
Another question, I'm just trying to understand your code decisions, any specific reason `scope`, `nodes` and `inject` are defined at the top of the script? 
Only `inject` is used outside the `linkjuice` const scope, and even then you could just pass it along in the `wrapNode` call.

**Comment by klaascuvelier** [comment]
I took a first try on this, let me know what you think. https://github.com/klaascuvelier/linkjuice/pull/1/files
(It's not a PR to your repo yet)

**Comment by klaascuvelier** [comment]
ping @toddmotto, did you see this?

**Comment by toddmotto** [comment]
@klaascuvelier Ahh, ja sorry maatje! Echt goed. Are there any edge cases you think we should be concerned of? If not, ping the PR here and we can test it and merge it :)

**Comment by klaascuvelier** [comment]
I tried this out myself with some different test cases and it worked well.

I just did another small change, now the TOC gets generated before the links are added. 
So TOC elements have original content (before running contentFn, so no icons, ...), and TOC links should work as the original elements should have an `id` already.

From my side this seems finished, let me know what you think @toddmotto
